src/httpserver.cpp

```diff
@@ -417,361 +286,708 @@ static void HTTPWorkQueueRun(WorkQueue<HTTPClosure>* queue, int worker_num)
+    auto minor = ToIntegral<int>(version_parts[1]);
+    if (!major || !minor) throw std::runtime_error("HTTP request line malformed");
+    m_version_major = major.value();
+    m_version_minor = minor.value();
 
-    g_work_queue = std::make_unique<WorkQueue<HTTPClosure>>(workQueueDepth);
-    // transfer ownership to eventBase/HTTP via .release()
-    eventBase = base_ctr.release();
-    eventHTTP = http_ctr.release();
     return true;
 }
 
-void UpdateHTTPServerLogging(bool enable) {
-    if (enable) {
-        event_enable_debug_logging(EVENT_DBG_ALL);
+bool HTTPRequest::LoadHeaders(LineReader& reader)
+{
+    return m_headers.Read(reader);
+}
+
+bool HTTPRequest::LoadBody(LineReader& reader)
+{
+    // https://httpwg.org/specs/rfc9112.html#message.body
+
+    auto transfer_encoding_header = m_headers.Find("Transfer-Encoding");
+    if (transfer_encoding_header && ToLower(transfer_encoding_header.value()) == "chunked") {
+        // Transfer-Encoding: https://datatracker.ietf.org/doc/html/rfc7230.html#section-3.3.1
+        // Chunked Transfer Coding: https://datatracker.ietf.org/doc/html/rfc7230.html#section-4.1
+        // see evhttp_handle_chunked_read() in libevent http.c
+        while (reader.Left() > 0) {
+            auto maybe_chunk_size = reader.ReadLine();
+            if (!maybe_chunk_size) return false;
+            uint64_t chunk_size;
+
+            if (!ParseUInt64Hex(maybe_chunk_size.value(), &chunk_size)) throw std::runtime_error("Invalid chunk size");
+
+            bool last_chunk{chunk_size == 0};
+
+            if (!last_chunk) {
+                // We are still expecting more data for this chunk
+                if (reader.Left() < chunk_size) {
+                    return false;
+                }
+                // Pack chunk onto body
+                m_body += reader.ReadLength(chunk_size);
+            }
+
+            // Even though every chunk size is explicitly declared,
+            // they are still terminated by a CRLF we don't need.
+            auto crlf = reader.ReadLine();
+            if (!crlf || crlf.value().size() != 0) throw std::runtime_error("Improperly terminated chunk");
+
+            if (last_chunk) return true;
+        }
+
+        // We read all the chunks but never got the last chunk, wait for client to send more
+        return false;
     } else {
-        event_enable_debug_logging(EVENT_DBG_NONE);
+        // No Content-length or Transfer-Encoding header means no body, see libevent evhttp_get_body()
+        auto content_length_value{m_headers.Find("Content-Length")};
+        if (!content_length_value) return true;
+
+        uint64_t content_length;
+        if (!ParseUInt64(content_length_value.value(), &content_length)) throw std::runtime_error("Cannot parse Content-Length value");
+
+        // Not enough data in buffer for expected body
+        if (reader.Left() < content_length) return false;
+
+        m_body = reader.ReadLength(content_length);
```

Comment:
If the body comes in pieces, then this parsing will be repeated for every piece and the `return` on line 460 will be executed a few times until enough data is received. This is suboptimal - to parse the same thing multiple times. I guess it is fine because we are not building a top performance HTTP server.

If this becomes an issue at some point, then it would be better to remember that it has been parsed until e.g. byte 456 and only continue parsing from there as new data is received over the socket. I guess this will complicate the current implementation significantly.